---
title: 'Postupy: Přijímání oznámení o první odpovídající výjimce'
description: Než modul CLR vyhledá obslužné rutiny výjimek, získáte oznámení o první pravděpodobnosti v rozhraní .NET prostřednictvím události FirstChanceException třídy AppDomain.
ms.date: 03/30/2017
dev_langs:
- csharp
- vb
helpviewer_keywords:
- first-chance exception notifications
- exceptions, first chance notifications
ms.assetid: 66f002b8-a97d-4a6e-a503-2cec01689113
ms.openlocfilehash: e8b5ae5fb69c7befd329316aee11523f79d73fcd
ms.sourcegitcommit: 1c37a894c923bea021a3cc38ce7cba946357bbe1
ms.translationtype: MT
ms.contentlocale: cs-CZ
ms.lasthandoff: 06/19/2020
ms.locfileid: "85104746"
---
# <a name="how-to-receive-first-chance-exception-notifications"></a><span data-ttu-id="5fd37-103">Postupy: Přijímání oznámení o první odpovídající výjimce</span><span class="sxs-lookup"><span data-stu-id="5fd37-103">How to: Receive First-Chance Exception Notifications</span></span>
<span data-ttu-id="5fd37-104"><xref:System.AppDomain.FirstChanceException>Událost <xref:System.AppDomain> třídy vám umožní obdržet oznámení o vyvolání výjimky, předtím než modul CLR (Common Language Runtime) začal hledat obslužné rutiny výjimek.</span><span class="sxs-lookup"><span data-stu-id="5fd37-104">The <xref:System.AppDomain.FirstChanceException> event of the <xref:System.AppDomain> class lets you receive a notification that an exception has been thrown, before the common language runtime has begun searching for exception handlers.</span></span>

 <span data-ttu-id="5fd37-105">Událost je vyvolána na úrovni domény aplikace.</span><span class="sxs-lookup"><span data-stu-id="5fd37-105">The event is raised at the application domain level.</span></span> <span data-ttu-id="5fd37-106">Vlákno provádění může projít více doménami aplikace, takže výjimka, která je neošetřena v jedné doméně aplikace, by mohla být zpracována v jiné doméně aplikace.</span><span class="sxs-lookup"><span data-stu-id="5fd37-106">A thread of execution can pass through multiple application domains, so an exception that is unhandled in one application domain could be handled in another application domain.</span></span> <span data-ttu-id="5fd37-107">K oznámení dojde v každé doméně aplikace, která přidala obslužnou rutinu pro událost, dokud aplikační doména nezpracovává výjimku.</span><span class="sxs-lookup"><span data-stu-id="5fd37-107">The notification occurs in each application domain that has added a handler for the event, until an application domain handles the exception.</span></span>

 <span data-ttu-id="5fd37-108">Postupy a příklady v tomto článku ukazují, jak dostávat oznámení o první pravděpodobné výjimce v jednoduchém programu, který má jednu doménu aplikace, a v doméně aplikace, kterou vytvoříte.</span><span class="sxs-lookup"><span data-stu-id="5fd37-108">The procedures and examples in this article show how to receive first-chance exception notifications in a simple program that has one application domain, and in an application domain that you create.</span></span>

 <span data-ttu-id="5fd37-109">Složitější příklad, který zahrnuje několik domén aplikace, najdete v příkladu pro <xref:System.AppDomain.FirstChanceException> událost.</span><span class="sxs-lookup"><span data-stu-id="5fd37-109">For a more complex example that spans several application domains, see the example for the <xref:System.AppDomain.FirstChanceException> event.</span></span>

## <a name="receiving-first-chance-exception-notifications-in-the-default-application-domain"></a><span data-ttu-id="5fd37-110">Přijímání oznámení o první odpovídající výjimce ve výchozí doméně aplikace</span><span class="sxs-lookup"><span data-stu-id="5fd37-110">Receiving First-Chance Exception Notifications in the Default Application Domain</span></span>
 <span data-ttu-id="5fd37-111">V následujícím postupu je vstupní bod pro aplikaci, `Main()` Metoda spouštěna ve výchozí doméně aplikace.</span><span class="sxs-lookup"><span data-stu-id="5fd37-111">In the following procedure, the entry point for the application, the `Main()` method, runs in the default application domain.</span></span>

#### <a name="to-demonstrate-first-chance-exception-notifications-in-the-default-application-domain"></a><span data-ttu-id="5fd37-112">Ukázka oznámení o první odpovídající výjimce ve výchozí doméně aplikace</span><span class="sxs-lookup"><span data-stu-id="5fd37-112">To demonstrate first-chance exception notifications in the default application domain</span></span>

1. <span data-ttu-id="5fd37-113">Definujte obslužnou rutinu události pro <xref:System.AppDomain.FirstChanceException> událost, pomocí funkce lambda a připojte ji k události.</span><span class="sxs-lookup"><span data-stu-id="5fd37-113">Define an event handler for the <xref:System.AppDomain.FirstChanceException> event, using a lambda function, and attach it to the event.</span></span> <span data-ttu-id="5fd37-114">V tomto příkladu obslužná rutina události vytiskne název domény aplikace, ve které byla událost zpracována, a <xref:System.Exception.Message%2A> vlastnost výjimky.</span><span class="sxs-lookup"><span data-stu-id="5fd37-114">In this example, the event handler prints the name of the application domain where the event was handled and the exception's <xref:System.Exception.Message%2A> property.</span></span>

     [!code-csharp[System.AppDomain.FirstChanceException_howto_simple#2](../../../samples/snippets/csharp/VS_Snippets_CLR_System/system.appdomain.firstchanceexception_howto_simple/cs/example.cs#2)]
     [!code-vb[System.AppDomain.FirstChanceException_howto_simple#2](../../../samples/snippets/visualbasic/VS_Snippets_CLR_System/system.appdomain.firstchanceexception_howto_simple/vb/example.vb#2)]

2. <span data-ttu-id="5fd37-115">Vyvolejte výjimku a zachyťte ji.</span><span class="sxs-lookup"><span data-stu-id="5fd37-115">Throw an exception and catch it.</span></span> <span data-ttu-id="5fd37-116">Předtím, než modul runtime vyhledá obslužnou rutinu výjimky, <xref:System.AppDomain.FirstChanceException> událost je vyvolána a zobrazí zprávu.</span><span class="sxs-lookup"><span data-stu-id="5fd37-116">Before the runtime locates the exception handler, the <xref:System.AppDomain.FirstChanceException> event is raised and displays a message.</span></span> <span data-ttu-id="5fd37-117">Tato zpráva je následována zprávou, která je zobrazena obslužnou rutinou výjimky.</span><span class="sxs-lookup"><span data-stu-id="5fd37-117">This message is followed by the message that is displayed by the exception handler.</span></span>

     [!code-csharp[System.AppDomain.FirstChanceException_howto_simple#3](../../../samples/snippets/csharp/VS_Snippets_CLR_System/system.appdomain.firstchanceexception_howto_simple/cs/example.cs#3)]
     [!code-vb[System.AppDomain.FirstChanceException_howto_simple#3](../../../samples/snippets/visualbasic/VS_Snippets_CLR_System/system.appdomain.firstchanceexception_howto_simple/vb/example.vb#3)]

3. <span data-ttu-id="5fd37-118">Vyvolejte výjimku, ale nezachyťte ji.</span><span class="sxs-lookup"><span data-stu-id="5fd37-118">Throw an exception, but do not catch it.</span></span> <span data-ttu-id="5fd37-119">Předtím, než modul runtime vyhledá obslužnou rutinu výjimky, <xref:System.AppDomain.FirstChanceException> událost je vyvolána a zobrazí zprávu.</span><span class="sxs-lookup"><span data-stu-id="5fd37-119">Before the runtime looks for an exception handler, the <xref:System.AppDomain.FirstChanceException> event is raised and displays a message.</span></span> <span data-ttu-id="5fd37-120">Není k dispozici obslužná rutina výjimky, takže aplikace se ukončí.</span><span class="sxs-lookup"><span data-stu-id="5fd37-120">There is no exception handler, so the application terminates.</span></span>

     [!code-csharp[System.AppDomain.FirstChanceException_howto_simple#4](../../../samples/snippets/csharp/VS_Snippets_CLR_System/system.appdomain.firstchanceexception_howto_simple/cs/example.cs#4)]
     [!code-vb[System.AppDomain.FirstChanceException_howto_simple#4](../../../samples/snippets/visualbasic/VS_Snippets_CLR_System/system.appdomain.firstchanceexception_howto_simple/vb/example.vb#4)]

     <span data-ttu-id="5fd37-121">Kód, který je zobrazen v prvním třech krocích tohoto postupu, tvoří úplnou konzolovou aplikaci.</span><span class="sxs-lookup"><span data-stu-id="5fd37-121">The code that is shown in the first three steps of this procedure forms a complete console application.</span></span> <span data-ttu-id="5fd37-122">Výstup z aplikace se liší v závislosti na názvu souboru. exe, protože název výchozí domény aplikace se skládá z názvu a přípony souboru. exe.</span><span class="sxs-lookup"><span data-stu-id="5fd37-122">The output from the application varies, depending on the name of the .exe file, because the name of the default application domain consists of the name and extension of the .exe file.</span></span> <span data-ttu-id="5fd37-123">Ukázku výstupu najdete v následujících tématech.</span><span class="sxs-lookup"><span data-stu-id="5fd37-123">See the following for sample output.</span></span>

     [!code-csharp[System.AppDomain.FirstChanceException_howto_simple#5](../../../samples/snippets/csharp/VS_Snippets_CLR_System/system.appdomain.firstchanceexception_howto_simple/cs/example.cs#5)]
     [!code-vb[System.AppDomain.FirstChanceException_howto_simple#5](../../../samples/snippets/visualbasic/VS_Snippets_CLR_System/system.appdomain.firstchanceexception_howto_simple/vb/example.vb#5)]

## <a name="receiving-first-chance-exception-notifications-in-another-application-domain"></a><span data-ttu-id="5fd37-124">Přijímání oznámení o první pravděpodobné výjimce v jiné doméně aplikace</span><span class="sxs-lookup"><span data-stu-id="5fd37-124">Receiving First-Chance Exception Notifications in Another Application Domain</span></span>
 <span data-ttu-id="5fd37-125">Pokud program obsahuje více než jednu doménu aplikace, můžete zvolit, které domény aplikace budou dostávat oznámení.</span><span class="sxs-lookup"><span data-stu-id="5fd37-125">If your program contains more than one application domain, you can choose which application domains receive notifications.</span></span>

#### <a name="to-receive-first-chance-exception-notifications-in-an-application-domain-that-you-create"></a><span data-ttu-id="5fd37-126">Příjem oznámení o první pravděpodobné výjimce v doméně aplikace, kterou vytvoříte</span><span class="sxs-lookup"><span data-stu-id="5fd37-126">To receive first-chance exception notifications in an application domain that you create</span></span>

1. <span data-ttu-id="5fd37-127">Definujte obslužnou rutinu události pro <xref:System.AppDomain.FirstChanceException> událost.</span><span class="sxs-lookup"><span data-stu-id="5fd37-127">Define an event handler for the <xref:System.AppDomain.FirstChanceException> event.</span></span> <span data-ttu-id="5fd37-128">V tomto příkladu se používá `static` metoda ( `Shared` metoda v Visual Basic), která vytiskne název domény aplikace, ve které byla událost zpracována, a <xref:System.Exception.Message%2A> vlastnost výjimky.</span><span class="sxs-lookup"><span data-stu-id="5fd37-128">This example uses a `static` method (`Shared` method in Visual Basic) that prints the name of the application domain where the event was handled and the exception's <xref:System.Exception.Message%2A> property.</span></span>

     [!code-csharp[System.AppDomain.FirstChanceException_howto#3](../../../samples/snippets/csharp/VS_Snippets_CLR_System/system.appdomain.firstchanceexception_howto/cs/example.cs#3)]
     [!code-vb[System.AppDomain.FirstChanceException_howto#3](../../../samples/snippets/visualbasic/VS_Snippets_CLR_System/system.appdomain.firstchanceexception_howto/vb/example.vb#3)]

2. <span data-ttu-id="5fd37-129">Vytvořte doménu aplikace a přidejte obslužnou rutinu události do <xref:System.AppDomain.FirstChanceException> události pro danou doménu aplikace.</span><span class="sxs-lookup"><span data-stu-id="5fd37-129">Create an application domain and add the event handler to the <xref:System.AppDomain.FirstChanceException> event for that application domain.</span></span> <span data-ttu-id="5fd37-130">V tomto příkladu je doména aplikace pojmenována `AD1` .</span><span class="sxs-lookup"><span data-stu-id="5fd37-130">In this example, the application domain is named `AD1`.</span></span>

     [!code-csharp[System.AppDomain.FirstChanceException_howto#2](../../../samples/snippets/csharp/VS_Snippets_CLR_System/system.appdomain.firstchanceexception_howto/cs/example.cs#2)]
     [!code-vb[System.AppDomain.FirstChanceException_howto#2](../../../samples/snippets/visualbasic/VS_Snippets_CLR_System/system.appdomain.firstchanceexception_howto/vb/example.vb#2)]

     <span data-ttu-id="5fd37-131">Tuto událost můžete zpracovat ve výchozí doméně aplikace stejným způsobem.</span><span class="sxs-lookup"><span data-stu-id="5fd37-131">You can handle this event in the default application domain in the same way.</span></span> <span data-ttu-id="5fd37-132">Použijte `static` vlastnost ( `Shared` v Visual Basic) <xref:System.AppDomain.CurrentDomain%2A?displayProperty=nameWithType> v aplikaci `Main()` , chcete-li získat odkaz na výchozí doménu aplikace.</span><span class="sxs-lookup"><span data-stu-id="5fd37-132">Use the `static` (`Shared` in Visual Basic) <xref:System.AppDomain.CurrentDomain%2A?displayProperty=nameWithType> property in `Main()` to get a reference to the default application domain.</span></span>

#### <a name="to-demonstrate-first-chance-exception-notifications-in-the-application-domain"></a><span data-ttu-id="5fd37-133">Ukázka oznámení o první pravděpodobnosti v doméně aplikace</span><span class="sxs-lookup"><span data-stu-id="5fd37-133">To demonstrate first-chance exception notifications in the application domain</span></span>

1. <span data-ttu-id="5fd37-134">Vytvořte `Worker` objekt v doméně aplikace, kterou jste vytvořili v předchozím postupu.</span><span class="sxs-lookup"><span data-stu-id="5fd37-134">Create a `Worker` object in the application domain that you created in the previous procedure.</span></span> <span data-ttu-id="5fd37-135">`Worker`Třída musí být veřejná a musí být odvozena z <xref:System.MarshalByRefObject> , jak je znázorněno v kompletním příkladu na konci tohoto článku.</span><span class="sxs-lookup"><span data-stu-id="5fd37-135">The `Worker` class must be public, and must derive from <xref:System.MarshalByRefObject>, as shown in the complete example at the end of this article.</span></span>

     [!code-csharp[System.AppDomain.FirstChanceException_howto#4](../../../samples/snippets/csharp/VS_Snippets_CLR_System/system.appdomain.firstchanceexception_howto/cs/example.cs#4)]
     [!code-vb[System.AppDomain.FirstChanceException_howto#4](../../../samples/snippets/visualbasic/VS_Snippets_CLR_System/system.appdomain.firstchanceexception_howto/vb/example.vb#4)]

2. <span data-ttu-id="5fd37-136">Zavolejte metodu `Worker` objektu, který vyvolá výjimku.</span><span class="sxs-lookup"><span data-stu-id="5fd37-136">Call a method of the `Worker` object that throws an exception.</span></span> <span data-ttu-id="5fd37-137">V tomto příkladu `Thrower` je metoda volána dvakrát.</span><span class="sxs-lookup"><span data-stu-id="5fd37-137">In this example, the `Thrower` method is called twice.</span></span> <span data-ttu-id="5fd37-138">Poprvé je argument metody `true` , což způsobí, že metoda zachytí svou vlastní výjimku.</span><span class="sxs-lookup"><span data-stu-id="5fd37-138">The first time, the method argument is `true`, which causes the method to catch its own exception.</span></span> <span data-ttu-id="5fd37-139">Druhý čas, argument je `false` a `Main()` metoda zachytí výjimku ve výchozí doméně aplikace.</span><span class="sxs-lookup"><span data-stu-id="5fd37-139">The second time, the argument is `false`, and the `Main()` method catches the exception in the default application domain.</span></span>

     [!code-csharp[System.AppDomain.FirstChanceException_howto#6](../../../samples/snippets/csharp/VS_Snippets_CLR_System/system.appdomain.firstchanceexception_howto/cs/example.cs#6)]
     [!code-vb[System.AppDomain.FirstChanceException_howto#6](../../../samples/snippets/visualbasic/VS_Snippets_CLR_System/system.appdomain.firstchanceexception_howto/vb/example.vb#6)]

3. <span data-ttu-id="5fd37-140">Umístěte kód do `Thrower` metody pro řízení, zda metoda zpracovává svou vlastní výjimku.</span><span class="sxs-lookup"><span data-stu-id="5fd37-140">Place code in the `Thrower` method to control whether the method handles its own exception.</span></span>

     [!code-csharp[System.AppDomain.FirstChanceException_howto#5](../../../samples/snippets/csharp/VS_Snippets_CLR_System/system.appdomain.firstchanceexception_howto/cs/example.cs#5)]
     [!code-vb[System.AppDomain.FirstChanceException_howto#5](../../../samples/snippets/visualbasic/VS_Snippets_CLR_System/system.appdomain.firstchanceexception_howto/vb/example.vb#5)]

## <a name="example"></a><span data-ttu-id="5fd37-141">Příklad</span><span class="sxs-lookup"><span data-stu-id="5fd37-141">Example</span></span>
 <span data-ttu-id="5fd37-142">Následující příklad vytvoří doménu aplikace s názvem `AD1` a přidá obslužnou rutinu události do události domény aplikace <xref:System.AppDomain.FirstChanceException> .</span><span class="sxs-lookup"><span data-stu-id="5fd37-142">The following example creates an application domain named `AD1` and adds an event handler to the application domain's <xref:System.AppDomain.FirstChanceException> event.</span></span> <span data-ttu-id="5fd37-143">Příklad vytvoří instanci `Worker` třídy v aplikační doméně a zavolá metodu s názvem `Thrower` , která vyvolá <xref:System.ArgumentException> .</span><span class="sxs-lookup"><span data-stu-id="5fd37-143">The example creates an instance of the `Worker` class in the application domain, and calls a method named `Thrower` that throws an <xref:System.ArgumentException>.</span></span> <span data-ttu-id="5fd37-144">V závislosti na hodnotě jeho argumentu Metoda buď zachytí výjimku, nebo ji nedokáže zpracovat.</span><span class="sxs-lookup"><span data-stu-id="5fd37-144">Depending on the value of its argument, the method either catches the exception or fails to handle it.</span></span>

 <span data-ttu-id="5fd37-145">Pokaždé, když `Thrower` metoda vyvolá výjimku v `AD1` , <xref:System.AppDomain.FirstChanceException> událost je vyvolána v `AD1` a obslužná rutina události zobrazí zprávu.</span><span class="sxs-lookup"><span data-stu-id="5fd37-145">Each time the `Thrower` method throws an exception in `AD1`, the <xref:System.AppDomain.FirstChanceException> event is raised in `AD1`, and the event handler displays a message.</span></span> <span data-ttu-id="5fd37-146">Modul runtime pak vyhledá obslužnou rutinu výjimky.</span><span class="sxs-lookup"><span data-stu-id="5fd37-146">The runtime then looks for an exception handler.</span></span> <span data-ttu-id="5fd37-147">V prvním případě je obslužná rutina výjimky nalezena v `AD1` .</span><span class="sxs-lookup"><span data-stu-id="5fd37-147">In the first case, the exception handler is found in `AD1`.</span></span> <span data-ttu-id="5fd37-148">Ve druhém případě je výjimka neošetřena v `AD1` a místo toho je zachycena ve výchozí doméně aplikace.</span><span class="sxs-lookup"><span data-stu-id="5fd37-148">In the second case, the exception is unhandled in `AD1`, and instead is caught in the default application domain.</span></span>

> [!NOTE]
> <span data-ttu-id="5fd37-149">Název výchozí domény aplikace je stejný jako název spustitelného souboru.</span><span class="sxs-lookup"><span data-stu-id="5fd37-149">The name of the default application domain is the same as the name of the executable.</span></span>

 <span data-ttu-id="5fd37-150">Pokud přidáte obslužnou rutinu <xref:System.AppDomain.FirstChanceException> události do výchozí domény aplikace, událost je vyvolána a zpracována před tím, než výchozí aplikační doména zpracuje výjimku.</span><span class="sxs-lookup"><span data-stu-id="5fd37-150">If you add a handler for the <xref:System.AppDomain.FirstChanceException> event to the default application domain, the event is raised and handled before the default application domain handles the exception.</span></span> <span data-ttu-id="5fd37-151">Chcete-li se podívat, přidejte kód jazyka C# `AppDomain.CurrentDomain.FirstChanceException += FirstChanceException;` (v Visual Basic, `AddHandler AppDomain.CurrentDomain.FirstChanceException, FirstChanceException` ) na začátku `Main()` .</span><span class="sxs-lookup"><span data-stu-id="5fd37-151">To see this, add the C# code `AppDomain.CurrentDomain.FirstChanceException += FirstChanceException;` (in Visual Basic, `AddHandler AppDomain.CurrentDomain.FirstChanceException, FirstChanceException`) at the beginning of `Main()`.</span></span>

 [!code-csharp[System.AppDomain.FirstChanceException_howto#1](../../../samples/snippets/csharp/VS_Snippets_CLR_System/system.appdomain.firstchanceexception_howto/cs/example.cs#1)]
 [!code-vb[System.AppDomain.FirstChanceException_howto#1](../../../samples/snippets/visualbasic/VS_Snippets_CLR_System/system.appdomain.firstchanceexception_howto/vb/example.vb#1)]

## <a name="see-also"></a><span data-ttu-id="5fd37-152">Viz také</span><span class="sxs-lookup"><span data-stu-id="5fd37-152">See also</span></span>

- <xref:System.AppDomain.FirstChanceException>
